src/unpack.c

```diff
@@ -189,20 +189,27 @@ static inline int template_callback_false(unpack_user* u, msgpack_object* o)
 
 static inline int template_callback_array(unpack_user* u, unsigned int n, msgpack_object* o)
 {
-    unsigned int size;
+	size_t size;
+    // Let's leverage the fact that sizeof(msgpack_object) is a compile time constant
+    // to check for int overflows.
+    // Note - while n is constrained to 32-bit, the product of n * sizeof(msgpack_object)
+    // might not be constrained to 4GB on 64-bit systems
+    if( n > SIZE_MAX/sizeof(msgpack_object))
```

Comment:
It seems that the comparison is always false on some 64bit environment.
See https://travis-ci.org/msgpack/msgpack-c/jobs/529103694#L1553
It is a warning but the CI regard it as error.

I don't come up with how to fix it, so far.

Reply:
I noticed `n` is `unsigned int`. So the solution is 
```c
if ((size_t)n > > SIZE_MAX/sizeof(msgpack_object))
```

Reply:
I don't know if that will help. Do we happen to detect somewhere whether we have 64 or 32-bit environment, perhaps set a #define that we can check. If the compiler is smart enough, it will sort this out. I'll try fixing the tab, and try this, we can see what it will do.

Reply:
```c
if ((size_t)n > SIZE_MAX/sizeof(msgpack_object))
```

Extra `>` is my typo. On typical 64bit environment, `SIZE_MAX` is `0xffffffffffffffff`, `sizeof(msgpack_object)` is `24`. The max value of `unsigned int` is `0xffffffff`.
so `0xffffffff > 0xffffffffffffffff / 24` is always false. Compiler warns it.

Reply:
Yes, that's exactly the problem. We'd prefer to only do this check in 32-bit. The issue is that doing it this way causes the check to only be a compile time constant comparison, which is very efficient, and what we want. But in 64-bit, it is too efficient, and the compiler complains instead of discarding the branch completely.

The size_t cast _might_ fool it, let's see. If this were C++, I can easily fix the problem with a template specialization, but this isn't C++ - though it does seem to compile cleanly as C++. I'm trying not to cause large, possibly disruptive changes.

I have another couple of angles I can try that will get us there, possibly an enum. Do we perhaps have a #defined constant that lets me know if it is 64-bit, and then I can just #ifdef around it?

Reply:
Also, the way the cast is defined, doing (size_t)n can only ever result in a value of 0x00000000fffffffff, because n is unsigned, and must zero extend. So maybe the cast will get the clever compiler to be quiet, but maybe not. Let's see what it does with it.

Reply:
Now, I'm watching travis-ci result (still working). The warnings on clang are disappeared. But gcc is still outputs warnings.

The operator `sizeof()` is evaluated compile time, not preprocess time, but macro `SIZE_MAX` is evaluated preprocess time. So something like the following code would help

```c
#if SIZE_MAX > UINT_MAX
```